microbench/src/main/java/io/netty/microbench/concurrent/DefaultPromiseBenchmark.java

```diff
@@ -0,0 +1,133 @@
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+@State(Scope.Benchmark)
+@Fork(1)
+@Threads(1)
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+public class DefaultPromiseBenchmark extends AbstractMicrobenchmark {
+    private EventExecutor executor;
+    private TestGenericFutureListener[] listeners;
+    private TestGenericFutureListener[] lastListeners;
+    private static AtomicLong doneCount;
+
+    @Param({ "1000000" })
+    public int numPromises;
+
+    @Param({ "1", "4" })
+    public int numListeners;
+
+    @Param({ "0", "4" })
+    public int lateListeners;
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class DoneCounter {
+        @Setup(Level.Iteration)
+        public void clean() {
+            doneCount.set(0);
+        }
+
+        public long messagesPerSecond() {
+            return doneCount.get();
+        }
+    }
+
+    @Setup(Level.Trial)
+    public void setup() {
+        doneCount = new AtomicLong();
+        executor = new DefaultEventExecutor();
+        listeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < numListeners; ++i) {
+            listeners[i] = new TestGenericFutureListener();
+        }
+        lastListeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < numListeners; ++i) {
+            lastListeners[i] = new TestGenericFutureListener();
+        }
+    }
+
+    @TearDown(Level.Trial)
+    public void teardown() {
+        executor.shutdownGracefully(0, 0, TimeUnit.MILLISECONDS);
+    }
+
+    @Benchmark
+    public void testInEventLoop(DoneCounter doneCounter) throws Exception {
+        CountDownLatch latch = new CountDownLatch(1);
+        startInEventLoop(latch);
+        latch.await();
+        Thread.sleep(1001);
+    }
+
+    @Benchmark
+    public void testOutEventLoop(DoneCounter doneCounter) throws Exception {
+        startOutEventLoop();
+        Thread.sleep(1001);
+    }
+
+    private void startInEventLoop(final CountDownLatch latch) {
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                latch.countDown();
+                startOutEventLoop();
+            }
+        });
+    }
+
+    private void startOutEventLoop() {
+        for (int i = 0; i < numPromises; ++i) {
+            Promise<Void> p = new DefaultPromise<Void>(executor);
```

Comment:
I think we need to be more clear about what we're testing here.  If we're really testing completion and listener performance, we probably want to avoid the allocation in the test.